src/pki/commands/ocsp.c

```diff
@@ -0,0 +1,569 @@
+				{
+					error = "invalid --digest type";
+					goto usage;
+				}
+				continue;
+			case 'R':
+				if (!parse_rsa_padding(arg, &pss))
+				{
+					error = "invalid RSA padding";
+					goto usage;
+				}
+				continue;
+			case EOF:
+				break;
+			default:
+				error = "invalid --ocsp option";
+				goto usage;
+		}
+		break;
+	}
+
+	lib->credmgr->add_local_set(lib->credmgr, &creds->set, FALSE);
+
+	responses = linked_list_create();
+
+	if (op == OP_RESPOND && !cacert)
+	{
+		error = "respond mode requires a ca certificate";
+		goto end;
+	}
+
+	if (op == OP_RESPOND && !key)
+	{
+		error = "respond mode requires a private signer key";
+		goto end;
+	}
+
+	/* re-initialize signing certificate and key pointers */
+	cert = NULL;
+	key = NULL;
+
+	if (file)
+	{
+		ocsp_req = lib->creds->create(lib->creds, CRED_CERTIFICATE,
+								CERT_X509_OCSP_REQUEST,
+								BUILD_FROM_FILE, file, BUILD_END);
+	}
+	else
+	{
+		chunk_t chunk;
+
+		set_file_mode(stdin, CERT_ASN1_DER);
+		if (!chunk_from_fd(0, &chunk))
+		{
+			fprintf(stderr, "%s: ", strerror(errno));
+			error = "reading certificate request failed";
+			goto end;
+		}
+		ocsp_req = lib->creds->create(lib->creds, CRED_CERTIFICATE,
+								CERT_X509_OCSP_REQUEST,
+								BUILD_BLOB_ASN1_DER, chunk, BUILD_END);
+		free(chunk.ptr);
+	}
+	if (!ocsp_req)
+	{
+		if (op == OP_SHOW)
+		{
+			error = "malformed OCSP request";
+			goto end;
+		}
+		else
+		{
+			DBG1(DBG_LIB, "malformed OCSP request");
+			ocsp_status = OCSP_MALFORMEDREQUEST;
+			goto gen;
+		}
+	}
+	ocsp_request = (ocsp_request_t*)ocsp_req;
+
+	/* does the requestor identify itself? */
+	requestor = ocsp_req->get_subject(ocsp_req);
+	if (requestor)
+	{
+		DBG1(DBG_LIB, "requestor:          \"%Y\"", requestor);
+
+		/* verify an optional ocsp request signature */
+		if (!verify_signature(ocsp_req, creds))
+		{
+			ocsp_status = OCSP_UNAUTHORIZED;
+			goto gen;
+		}
+	}
+
+	/* extract nonce from OCSP request */
+	nonce = ocsp_request->get_nonce(ocsp_request);
+	if (nonce.len > 0)
+	{
+		DBG1(DBG_LIB, "nonce:               %#B", &nonce);
+	}
+
+	/* check for an ocsp responder */
+	if (op == OP_RESPOND)
+	{
+		ocsp_responder = lib->get(lib, "ocsp-responder");
+		if (!ocsp_responder)
+		{
+			DBG1(DBG_LIB, " no ocsp-responder found");
+			ocsp_status = OCSP_INTERNALERROR;
+			goto gen;
+		}
+	}
+
+	/* enumerate over the ocsp requests and try to identify the issuers */
+	enumerator = ocsp_request->create_request_enumerator(ocsp_request);
+	while (enumerator->enumerate(enumerator, &hashAlgorithm, &issuerNameHash,
+											 &issuerKeyHash, &serialNumber))
+	{
+		certificate_t *issuer_cacert = NULL;
+		cert_validation_t status = VALIDATION_FAILED;
+		ocsp_single_response_t *response = NULL;
+		crl_reason_t revocationReason;
+		time_t revocationTime;
+
+		/* search for the matching issuer cacert */
+		if (find_issuer_cacert(hashAlgorithm, issuerKeyHash, issuerNameHash,
+							   &issuer_cacert))
+		{
+			if (first_issuer == NULL)
```

Comment:
```suggestion
			if (!first_issuer)
```

Reply:
These are personal preferences, but in order to prevent a third world war, I changed the statement